File: modules/expo-secure-credentials/android/src/main/java/expo/modules/securecredentials/ExpoSecureCredentialsModule.kt
package expo.modules.securecredentials

import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoSecureCredentialsModule : Module() {
  private fun getContext(): Context {
    return appContext.reactContext ?: throw Error("React context is null")
  }

  // Lazy-initialize to avoid re-deriving MasterKey on every bridge call.
  // EncryptedSharedPreferences.create() is expensive (crypto operations).
  private val storage: EncryptedStorage by lazy {
    EncryptedStorage(getContext())
  }

  override fun definition() =
    ModuleDefinition {
      Name("ExpoSecureCredentials")

      Function("canUseSecureStorage") {
        return@Function true
      }

      AsyncFunction("setItemAsync") { key: String, value: String ->
        storage.set(key, value)
      }

      AsyncFunction("getItemAsync") { key: String ->
        return@AsyncFunction storage.get(key)
      }

      AsyncFunction("deleteItemAsync") { key: String ->
        storage.delete(key)
      }
    }
}

File: modules/expo-secure-credentials/ios/ExpoSecureCredentialsModule.swift
import Foundation
import ExpoModulesCore

public class ExpoSecureCredentialsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoSecureCredentials")

    Function("canUseSecureStorage") {
      return true
    }

    AsyncFunction("setItemAsync") { (key: String, value: String) in
      try KeychainHelper.shared.set(value, forKey: key)
    }

    AsyncFunction("getItemAsync") { (key: String) -> String? in
      return try KeychainHelper.shared.get(forKey: key)
    }

    AsyncFunction("deleteItemAsync") { (key: String) in
      KeychainHelper.shared.delete(forKey: key)
    }
  }
}
